### PR TITLE
PostgreSQL: add initial support (WIP)

### PR DESCRIPTION
**Pull request by jufajardini**
Make sure these boxes are signed before submitting your Pull Request -- thank you.

- [x] I have read the contributing guidelines at https://redmine.openinfosecfoundation.org/projects/suricata/wiki/Contributing
- [x] I have signed the Open Information Security Foundation contribution agreement at https://suricata-ids.org/about/contribution-agreement/
- [ ] I have updated the user guide (in doc/userguide/) to reflect the changes made (if applicable)

Link to [redmine](https://redmine.openinfosecfoundation.org/projects/suricata/issues) ticket:
https://redmine.openinfosecfoundation.org/issues/4241

Previous PR:
https://github.com/OISF/suricata/pull/6013

Describe changes:
rust/src/applayerpgsql/parser.rs:
- add nom parsers for decoding most messages involved in Startup Phase
- add more unittests
src/Makefile.am fix formatting issue
rust/src/applayerpgsql/pgsql.rs:
- fix unused documentation comments warning
output-json-pgsql.c:
- update CreateEveHeader call.

PR description: of the messages that compose PostgreSQL Startup Phase, this decoder can parse most of them, except for:
- `NegotiateProtocolVersion`
- `GSSAPIAuthentication` messages
- `AuthenticationKerberosV5'
The logger needs work, so it will offer filterable fields
Some of the messages that are parsed are not integrated into the decoder, for they are dependent of events handling, and this parser doesn't cover that, yet. (e.g. SSL Handshake - STARTTLS; Backend Initialization Messages, which happen after successful Authentication)

Current state of things:
- I believe the next steps/ tasks are related to implementing events, in order to be able to handle the aforementioned situations. I have written a proposal for such events, explaining the reasoning behind them, here: https://docs.google.com/document/d/1FRkKfAL741l_tnRhudj3Rx3LJbbo9WTW3M4bbLlaiw8/edit?usp=sharing
- With that working, then I think I could move on to the simple query subprotocol, to have a v1 of the redmine task

#suricata-verify-pr:
#suricata-verify-repo: 
#suricata-verify-branch: https://github.com/jufajardini/suricata-verify/tree/pgsql-startup-phase-test-v0 
#suricata-update-pr:
#suricata-update-repo:
#suricata-update-branch:
#libhtp-pr:
#libhtp-repo:
#libhtp-branch:

### CODE REVIEW

**Comment by jufajardini** [comment]
Confused about the CIFuzz faillures. Did I do anything wrong that I'm not aware of?

**Comment by jufajardini** [comment]
While I studied events implementations in the codebase today, I realized that I'm likely misplacing events for Transaction's state. I think this makes sense, based on previous answers, so, while I'm still waiting for some feedback, I'll re-check what I have on Transaction's state, and get a better understanding of how Rust parsers handle those, in the hopes that that may help me understand what exactly is that I should use...

**Comment by catenacyber** [comment]
> Confused about the CIFuzz failures. Did I do anything wrong that I'm not aware of?

No, oss-fuzz was broken.
You can see a real failure if it happens during the `Run fuzzers` stage, not the `Build fuzzers` stage

File: rust/src/applayerpgsql/logger.rs [hunk 1]
```diff
@@ -0,0 +1,36 @@
+/* Copyright (C) 2018 Open Information Security Foundation
```
Comment (catenacyber) [inline rust/src/applayerpgsql/logger.rs:1]: Should the directory be named simply pqsql ?

Comment (catenacyber) [reply rust/src/applayerpgsql/logger.rs:1]: There should also be update in doc/userguide

File: rust/src/applayerpgsql/pgsql.rs [hunk 2]
```diff
@@ -0,0 +1,699 @@
...
+                    // of how much data is missing so just ask for one more byte so the
+                    // parse is called as soon as more data is received.
+                    SCLogNotice!("Suricata thinks request is incomplete");
+                    let consumed = input.len() - start.len();
+                    let needed_estimation = start.len() + 1;
+                    SCLogNotice!("Needed: {:?}, estimated needed: {:?}", needed, needed_estimation);
+                    return AppLayerResult::incomplete(consumed as u32, needed_estimation as u32);
+                },
+                Err(_) => {
+                    return AppLayerResult::err();
```
Comment (catenacyber) [inline rust/src/applayerpgsql/pgsql.rs:169]: Should we set an event in this case ? and have decoder-event rules in the `rules` directory ?

cf `git grep -i mqtt | cut -d: -f1 | uniq` to see which files need a change for a new app-layer protocol

File: rust/src/applayerpgsql/logger.rs (hunk 1 above)
Comment (catenacyber) [reply rust/src/applayerpgsql/logger.rs:1]: Is there no detection yet ? just logging ?

File: src/output.c [hunk 3]
```diff
@@ -76,6 +76,7 @@
 #include "output-json-sip.h"
```
Comment (catenacyber) [inline src/output.c:76]: `output-json-alert.c` should also be modified

File: suricata.yaml.in [hunk 4]
```diff
@@ -714,6 +715,8 @@ pcap-file:
 # "detection-only" enables protocol detection only (parser disabled).
 app-layer:
   protocols:
+    postgresql:
+      enabled: yes
```
Comment (catenacyber) [inline suricata.yaml.in:719]: Should it be enabled by default ?

File: src/suricata-common.h [hunk 5]
```diff
@@ -459,6 +459,7 @@ typedef enum {
     LOGGER_JSON_TEMPLATE_RUST,
```
Comment (catenacyber) [inline src/suricata-common.h:459]: `src/tests/fuzz/confyaml.c` should also be modified

**Review by catenacyber — COMMENTED** [review]
Some quick comments, hoping they are helpful to you

File: rust/src/applayerpgsql/logger.rs (hunk 1 above)
Comment (jufajardini) [reply rust/src/applayerpgsql/logger.rs:1]: That's right, no detection yet, just logging, a simplified version, for now. 
Indeed, I think the directory can be changed, I'll do that. And I'll also add the doc/userguide to my todo list!

File: rust/src/applayerpgsql/pgsql.rs (hunk 2 above)
Comment (jufajardini) [reply rust/src/applayerpgsql/pgsql.rs:169]: like `rules/mqtt-events.rules` ?

File: suricata.yaml.in (hunk 4 above)
Comment (jufajardini) [reply suricata.yaml.in:719]: Likely not, you're right.

File: src/suricata-common.h (hunk 5 above)
Comment (jufajardini) [reply src/suricata-common.h:459]: Should something about this also be added to the setup-app-layer.py script? What do you think, @jasonish ?

Comment (jasonish) [reply src/suricata-common.h:459]: I wasn't aware of this file.  @catenacyber What conditions would require this file to be updated? The additional/remove of any parser?

Comment (catenacyber) [reply src/suricata-common.h:459]: > I wasn't aware of this file. @catenacyber What conditions would require this file to be updated? The additional/remove of any parser?

Indeed.
I guess we could replace this part of the config by a util function like EnableAllProtocols...

Comment (catenacyber) [reply src/suricata-common.h:459]: PS : I was not aware of setup-app-layer.py ;-p

Comment (jufajardini) [reply src/suricata-common.h:459]: Hey @catenacyber, is adding the lines
```c
postgresql:
    enabled: yes
```
enough, or should I edit something else as well?

**Comment by suricata-qa** [comment]
| field | test | baseline | % diff |
|--|--|--|--|
| | tlpr1_stats_chk |
| .flow.spare | 6929064 | 5890991 | 85.0% |
| .flow.memuse | 1279679808 | 1130221888 | 88.0% |
| .app_layer.flow.ftp-data | 153 | 500 | 31.0% |

**Comment by jufajardini** [comment]
replaced by: https://github.com/OISF/suricata/pull/6281

**Pull request closed without merging**